WIP: Reworked Biome Block Counting for Vanilla Biomes

PaulHause opened the pull request:
### What is the new feature?

This is an attempt to partially address the issue discussed in [Issue#1339](https://github.com/tModLoader/tModLoader/issues/1399).

This is a slight rework of the way Terraria handles counting biome related blocks with concern to vanilla biomes. Instead of using the hardcoded tile numbers, the newly added counting mechanism uses the tile ID sets to allow for modders to add their own tiles to said sets and easily make a **modded** block count as part of a **vanilla** biome.

### Why should this be part of tModLoader?

This small addition allows modders to more easily make their own blocks count as part of vanilla biomes. If a modder wished to make a new stone type of block that counted to the game as being a part of the Hallow biome, for example, it was previously difficult to do so (perhaps even impossible). 

### Are there alternative designs?

This is where I ask for feedback. This is my first time attempting contribution to a large-scale project such as tModLoader and I would love pointers and feedback from anyone more familiar with the codebase. While very simple in approach, the current way I do this feels dangerous to me, as it essentially loops through every single tile in the game for every single frame of gameplay and it feels this could get bloated with how many tiles are added by a large number of mods. 

I would also appreciate any advice on if my code feels logical in its placement and usage, as well as my adherence to the style guide, as I had a larger patch than actual changes due to trying to keep my code legible to at least myself. There may very well be a better way to do this, but I wanted to at least get the ball rolling as this is a feature of interest to me for my own modding endeavors. (I am on the Discord server if you wish to chat with my directly as well: Username is Golden Guardian). 

### Sample usage for the new feature

The example I used above is good, but one I created for the purposes of testing my changes was to make a block that counted as both Snow and Hallow. This was to test if this implementation did not mess up any of vanilla Terraria's usual biome considerations. The expected result is that a small amount of the block would count as a Hallowed biome, and a much larger amount would count as both Hallowed and Snow. 

This was in fact the result and my own usage of this would be to further expand a mod allowing for Snow to be a tile that the Crimson, Corruption, and Hallow would spread to, not just ice. This could also allow modders to create entirely new structures within biomes constructed entirely of mod tiles without fearing overwriting the biome.

Chicken-Bones committed:
Decompile Terraria 1.4.4.1

Chicken-Bones committed:
1.4.4.1 Terraria patch step

Chicken-Bones committed:
1.4.4.1 TerrariaNetCore patch step

Chicken-Bones committed:
Implement NativeLibrary.TryLoad fixes for Razer SDK at TerrariaNetCore step

Chicken-Bones committed:
Fix NetCore server by passing headless param to Game base ctor

Chicken-Bones committed:
Delete/Rename patches for filename changes in 1.4.4

Mirsario committed:
1.4.4.2 Terraria & TerrariaNetCore patches

Mirsario committed:
1.4.4.2: Reworked Tile patches.

Mirsario committed:
1.4.4.2: Handle sound additions.

Mirsario committed:
1.4.4.2: New SoundUpdateCallback PlaySound parameter.

Chicken-Bones committed:
1.4.4.2: Attempt reimplementation of WaterStyleLoader DrawWater hook

Chicken-Bones committed:
1.4.4.2: Boss bar now has separate life+lifeMax (similar for shield)

Chicken-Bones committed:
1.4.4.2: Fix implementation of PlayerIO.Load and LoadPlayer_LastMinuteFixes

Chicken-Bones committed:
1.4.4.2: Preserve mod data in Player.SerializedClone

Chicken-Bones committed:
1.4.4.2: Add new music ID for shimmer

Chicken-Bones committed:
1.4.4.2: Move FinishPlayWorld into QueueMainThreadAction

- Fixes the player not actually spawning into the world, and addresses TODO

Chicken-Bones committed:
1.4.4.2 tModLoader patches

Co-authored-by: Mirsario <me@mirsar.io>

Chicken-Bones committed:
1.4.4.2: Initial compilation fixes

Co-authored-by: Mirsario <me@mirsar.io>

Chicken-Bones committed:
1.4.4.2: tModPorter updates, applied to ExampleMod

Chicken-Bones committed:
1.4.4.2: Address tModPorter suggestions in ExampleMod. Rework MinionBossBody

to show minion health as shields text.

Mirsario committed:
1.4.4.2: HookGen.

Mirsario committed:
1.4.4.2: Fixed a freeze in GetHurtByOtherNPCs.

Chicken-Bones committed:
1.4.4.2: Fix mod water styles not having a slope texture.

Chicken-Bones committed:
1.4.4.2: Fix crash with mod NPCs due to new array not being resized

Chicken-Bones committed:
1.4.4.2: Reimplement prefix hooks

Chicken-Bones committed:
1.4.4.2: Fix Utils.OpenFolder on windows

Chicken-Bones committed:
1.4.4.2: Preserve modded data when refreshing items (and reforging)

Chicken-Bones committed:
Update patches to 1.4.4.7

Chicken-Bones committed:
1.4.4.7: Remove TileLoader.MineDamage

- It had a bug which increased mining speed by 20%. Use power / modTile.MineResist directly instead.

Chicken-Bones committed:
1.4.4.7: Add tModPorter for TileID.Sets.TouchDamage changes

Chicken-Bones committed:
1.4.4.7: Add new player draw layers

Chicken-Bones committed:
1.4.4.7: Remove attack speed caps, and fix summon weapon speed patches

Chicken-Bones committed:
1.4.4.7: Cleanup bag opening hooks

JavidPack committed:
Add working directory to setup

JavidPack committed:
1.4.4.7: ExtractinatorUse changes, reimplementation, porter, and examples (#3085)

* ExtractinatorUse changes, reimplementation, porter, and examples

* Update PortingNotes_1.4.4.7.md

Chicken-Bones committed:
1.4.4.7: Use CommonMaxStack for all example items

Chicken-Bones committed:
1.4.4.7: Merge tML Item OnCreate hooks into new vanilla OnCreated

Chicken-Bones committed:
1.4.4.7: Better positioning of CanRightClick hook for openable items

Chicken-Bones committed:
tModPorter: Fix PlaySound rewriter with static Count field

Chicken-Bones committed:
tModPorter: Add not-yet-implemented markers so current tests can pass

Chicken-Bones committed:
Update TagIOTests so they all pass again

Chicken-Bones committed:
Update Roslyn to 4.3.1

- Decompiler dll was edited with dnSpy to fix breaking Humanizer version change

Chicken-Bones committed:
Switch to file scoped namespaces.

Chicken-Bones committed:
Fix collection initializer formatting

Mirsario committed:
Decompiler Update (#3083)

* Updated decompiler to '8.0.0.7106-preview2'.

* 'Terraria' patchstep.

* 'TerrariaNetCore' patchstep.

* 'tModLoader' patchstep.

* Editorconfig: Sort system directives first.

Mirsario committed:
Fixed vanilla step via a lost patch.

Mirsario committed:
Fixed offset of a single patch.

Chicken-Bones committed:
Remove old file from vanilla patch step

Mirsario committed:
File-scoped namespaces everywhere...

Mirsario committed:
Code Style Update (K&R, etc.) (#3097)

* Format Setup csproj files.

(cherry picked from commit 853f9dda5b4a64e68631ec45fd85d469b1fb1001)

* Improved csproj/targets generation with tabs & more.

* Switch to real K&R style - newlines on method declarations.

* Fixed indentation in case blocks

(cherry picked from commit de9a6fbdd6407df8966ad104bb2676f15a89838e)

* Re-undid a bad csproj change.

* 'Terraria' patchstep.

* TerrariaNetCore patchstep.

* 'tModLoader' patchstep done.

* Convert all missed methods, except non-patch TML files.

* Convert all remaining TML methods to K&R.

JavidPack committed:
1.4.4.7: DefaultToX usage in ExampleMod, Documentation (Take 2) #3088 (#3098)

* Reimplement code changes from #3088

* More fixes

Mirsario committed:
Merge branch '1.4' into 1.4.4

Mirsario committed:
Fixed crashes when deleting worlds/characters (Half of #3082)

Chicken-Bones committed:
Delete release_extras/Content folder from 1.4 merge

Chicken-Bones committed:
Delete old 1.4.3 porting notes

JavidPack committed:
1.4.4.7: Recipe shimmer examples and fixes (#3086)

* Shimmer recipe work, examples, comments, cloning

* Address feedback, move examples to ShimmerShowcase.cs

* Typo

* File-scoped namespace for the new file.

* Restore internalizing recipe bools

Co-authored-by: Mirsario <me@mirsar.io>

Chicken-Bones committed:
1.4.4.7: Remove some outdated NPC profile related code.

Chicken-Bones committed:
Fix failing tModPorter test after 1.4 merge

Chicken-Bones committed:
Remove nullable bool? from Mod/GlobalNPC.CanHitNPC

Chicken-Bones committed:
1.4.4.7: Update WorldGen.Convert documentation

Chicken-Bones committed:
1.4.4.7: Update/Rename hook, ModNPC.ApplyDifficultyAndPlayerScaling

Chicken-Bones committed:
Remove CanBurnInLava hook, as suitable replacements exist.

Chicken-Bones committed:
Change HookGen namespace style from IL/On. to IL/On_ on the type name

Chicken-Bones committed:
1.4.4.7: Add tModPorter comment to ApplyDifficultyAndPlayerScaling

Chicken-Bones committed:
Forgot to commit latest version of TerrariaHooks

Chicken-Bones committed:
Refactor a bunch of hook callsites into CombinedHooks

direwolf420 committed:
Fix tile framing (causing infinite UnloadedXTile loop and invisible tiles) (#3107)

* tile frame fix

* Minor fix to formatting after merge conflict fix

* Missed a line removal in merged conflict resolution

Co-authored-by: Solxanich <59670736+Solxanich@users.noreply.github.com>

Mirsario committed:
Patch improvements (#3105)

* Insane full patch review.

* Undo accidental non-cs reverts.

* Fixed asset (& mod icon) issues.

* Undo useless change.

* CB review of Terraria/TerrariaNetCore & tML ReLogic patches

* Strip UTF8 BOM from cs files

* CB review of tML patches

Co-authored-by: Chicken-Bones <mehvids@gmail.com>

Chicken-Bones committed:
1.4.4.9: Vanilla UIElement member refactor

Chicken-Bones committed:
Update patches to 1.4.4.9

Chicken-Bones committed:
Merge remote-tracking branch 'origin/1.4' into 1.4.4

direwolf420 committed:
[1.4.4] Change return type of Mod.AddContent to bool, add docs (#2953)

* change return type of Mod.AddContent to bool, add docs

* invert to avoid branching

* existing doc unification/clarification

* inheritdoc usage to capture doc changes

* fix bad merge

Chicken-Bones committed:
1.4.4: Fix rendering of Mod Configuration ingame menu option

Ved-s committed:
Ru localization (#3108)

* Ru localization

* Apply suggestions from code review

Co-authored-by: Mirsario <me@mirsar.io>

Solxanich committed:
Fixes Mod Tree Sapling placement (includes axe of growth) (#3118)

* Fixes Mod Tree Sapling placement (includes axe of growth)

* Update Porting Notes

Rijam committed:
[1.4.4] Obsidian Armor Set Bonus Whip Speed Fix (#3121)

* Obsidian armor whip speed fix

* Alternate fix

JavidPack committed:
Typo in ModifyHardmodeTasks docs

Jadams505 committed:
Fix layer drawing (#3124)

direwolf420 committed:
[1.4.4] Boss health bar update (#3117)

* Boss Health Bar updates
* Implement showText functionality
* Update comments for proper formatting
* Make public DrawHealthText method

* move DrawHealthText into partial

* misc ExampleMod updates

* comment

Mirsario committed:
Change DisplayNameInternal properties to 'private protected'.

Mirsario committed:
Workaround for window maximizing issues (Fixes #3082)

Mirsario committed:
Move & rename ModItem.SacrificeTotal (#3078)

* Moved ModItem.SacrificeTotal to Item

* ...Actually do what the previous commit said.

* Defense for the getter.

* Rename to ResearchUnlockCount.

* Minimize using changes.

* Leftovers.

* Add tModPorter refactor for SacrificeTotal -> Item.ResearchUnlockCount

Co-authored-by: Chicken-Bones <mehvids@gmail.com>

direwolf420 committed:
fix PrefixID.Sets not being resized (#3131)

direwolf420 committed:
more ModBiome/SceneEffect docs, changes (#3135)

direwolf420 committed:
[1.4.4] Log worldgen seed & consistency with existing world loading messages (#3130)

* log worldgen seed & consistency with existing world loading messages

* consts

* fix last Size=, change IsExpert to GameMode

Jadams505 committed:
Update vanillaSets with 1.4.4 options (#3136)

Jadams505 committed:
Visual chest transfer fix (#3129)

Sneakilli committed:
Added catch conditions for ExampleQuestFish (#3089)

* Added catch conditions for ExampleQuestFish

This fish could previously not be caught

* Update ExampleMod/Common/Players/ExampleFishingPlayer.cs

Co-authored-by: direwolf420 <amakejk1@web.de>

* Removed extraneous checks when catching ExampleQuestFish

Co-authored-by: direwolf420 <amakejk1@web.de>

Cyrillya committed:
[1.4.4] Infinite shooting fix (#3139)

* Set `itemAnimationMax` to -1 when entering world

* check `context == PlayerSpawnContext.SpawningIntoWorld`

* add an `itemAnimation > 0` && check to `ItemAnimationJustStarted`

* Move -1 comment into right spot

Co-authored-by: Chicken-Bones <mehvids@gmail.com>

Jadams505 committed:
[1.4.4] Remix seed fix (#3146)

* Remix seed generation fix

* Clear tile data

Cyrillya committed:
Fix quick heal & mana consuming non-consumable items (#3063)

* Add `item.consumable` check

* update 1.4.4

Cyrillya committed:
Call `ItemLoader.UseItem` for food quick buff (#3151)

Solxanich committed:
Workshop Improvements Pack 3 (#3157)

* Workshop fixes such as download refresh and futrther isolating publishing copy issues

* Cleanup comments and using statement

* Remove controversial preload browser data

Chicken-Bones committed:
Delete old 1.4.4.7 porting notes

PaulHause committed:
- Added new sets to TileID.TML.cs to facilitate easier block counting including modded blocks

- Created a RecountTiles function in TileLoader.cs to actually do the recounting using sets.

- Added the aformentioned function into SceneMetrics.cs to do the recounting.

- Added 2 lines to ensure the counts for Crimson and Corruption took sunflowers into account to maintain vanilla calculations.

Chicken-Bones committed:
Fix crash when missing the old local Content/Images override directory

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TileLoader.cs:
```diff
@@ -957,5 +957,34 @@ public static class TileLoader
+			metrics.HolyTileCount = metrics.EvilTileCount = metrics.BloodTileCount = metrics.SnowTileCount = metrics.JungleTileCount = metrics.MushroomTileCount = metrics.SandTileCount = 0;
+
+			for (int i = 0; i < TileCount; i++) {
+				if (TileID.Sets.HallowBiome[i] == true)
```
No need for `== true`

Chicken-Bones commented on patches/tModLoader/Terraria/SceneMetrics.cs.patch:
```diff
@@ -177,6 +177,18 @@
+ 			GraveyardTileCount = _tileCounts[85];
+ 			GraveyardTileCount -= _tileCounts[27] / 2;
++
++			TileLoader.RecountTiles(this);
```
Better to comment out the above code, if it's going to be ignored anyway

Chicken-Bones replied:
This currently doesn't handle the `Main.remixWorld` condition changes which tiles count as jungle/evil/blood

PaulHause replied:
Oh, I was unaware of this condition. I'm not seeing it in the source code I have, is this in a newer version? (I forked from 1.4 default, if that helps anything)

Chicken-Bones replied:
Ah, my apologies, I missed that this isn't targeting 1.4.4.

Development for 1.4 has stopped, so please rebase this onto the 1.4.4 branch.

PaulHause replied:
Understood, I'll do that. Should be a simple enough switch.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TileLoader.cs:
```diff
@@ -957,5 +957,34 @@ public static class TileLoader
+			// reset every tile count
+			metrics.HolyTileCount = metrics.EvilTileCount = metrics.BloodTileCount = metrics.SnowTileCount = metrics.JungleTileCount = metrics.MushroomTileCount = metrics.SandTileCount = 0;
+
+			for (int i = 0; i < TileCount; i++) {
```
Should store the result of `metrics._tileCounts[i]` in a local variable, and `continue` the loop if it's 0.

PaulHause replied:
For my own curiosity, any reason to store it in a local variable rather than making a direct check? This is what I do in my most recent commit. Is this considered bad practice?

Chicken-Bones replied:
Generally a good idea not to duplicate code (Terraria is a very poor example of this practice). Looking up an array repeatedly costs more than doing it once and saving the result.

Chicken-Bones replied:
Also same goes for below, if you're going to check `TileID.Sets.HallowBiome[i] != 0` you may as well just do the multiplication and add directly.

PaulHause replied:
OH, I understand what you mean now. I somehow entirely overlooked that I use it for every line that it is adding to the counts even though that is the entire point, ahah.

I can fix that right away.

Chicken-Bones commented on patches/tModLoader/Terraria/ID/TileID.TML.cs:
```diff
@@ -69,6 +69,17 @@ partial class Sets
+			/// <summary>
+			/// New created sets to facilitate vanilla biome block counting including modded blocks. To replace the current hardcoded counts in SceneMetrics.cs
+			/// </summary>
+			public static bool[] CorruptBiome = Factory.CreateBoolSet(23, 24, 25, 32, 112, 163, 400, 398);
```
Suggest, using `IntSet` instead, and we could put the 'score' values in here (1 normally, but it would support when a tile counts for multiple, or negative)
Eg `-10` evil for sunflowers

Chicken-Bones reviewed (changes requested):
Thankyou for your first contribution, please see my feedback.

PaulHause committed:
- Commented out unneccesary code.

- Modified biome tile sets to be an int set rather than a bool to allow values/strengths to be applied on a per tile basis.

- Added a continue condition to the counting loop to help speed up execution.

Thanks to Chicken-Bones for the suggestions.

PaulHause commented:
This is going to look far too messy as I fix my own repository to rebase to 1.4.4, going to close this and make a new request when I've rebased correctly.

Chicken-Bones commented:
Force push is always an option!

On Tue, 3 Jan 2023, 2:38 pm Golden Guardian, ***@***.***>
wrote:

> This is going to look far too messy as I fix my own repository to rebase
> to 1.4.4, going to close this and make a new request when I've rebased
> correctly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tModLoader/tModLoader/pull/3166#issuecomment-1369388905>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC6ZCNQXIIEGQNIOBA4HDDWQOUKRANCNFSM6AAAAAATPE4DVY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
